rfcs/proposed/task_arena_waiting/readme.md

```diff
@@ -0,0 +1,194 @@
+ta.wait_for( tg );              // => ta.execute([&tg]{ tg.wait(); });
+```
+If justified performance-wise, a more elaborated implementation could perhaps shave off some overheads.
+It would likely require new library entry points, though.
+
+The example code to split work across NUMA-bound task arenas could then look like this (assuming also
+a special function that creates and initializes a vector of arenas):
+```cpp
+std::vector<tbb::task_arena> numa_arenas =
+    initialize_constrained_arenas(/*some arguments*/);
+std::vector<tbb::task_group> task_groups(numa_arenas.size());
+
+for(unsigned j = 0; j < numa_arenas.size(); j++) {
+    numa_arenas[j].enqueue( (){/*some parallel stuff*/}, task_groups[j] );
+}
+
+for(unsigned j = 0; j < numa_arenas.size(); j++) {
+    numa_arenas[j].wait_for( task_groups[j] );
+}
+```
+
+Additionally, a task group parameter can also be added to `execute` as well as to `this_task_arena::isolate`,
+which is similar to `execute` but provides additional work isolation guarantees. That would make
+the experimental `isolated_task_group` class fully obsolete.
+
+### 2. Reconsider waiting for all tasks
+
+With the redesign of the task scheduler in oneTBB, we might evaluate if there is now a reliable way
+to ensure completion of all jobs in an arena. The safety concern can potentially be mitigated
+by throwing an exception or returning `false`, similarly to the `tbb::finalize` function.
+
+The corresponding API could look like:
+```cpp
+ta.wait_for_all(); // throws tbb::unsafe_wait if called 
+bool success = ta.wait_for_all(std::nothrow{}); // in case exceptions are undesirable
+```
+Another option for the non-throwing variation is `bool try_wait_for_all()`. With that name, it might also
+have a weaker semantics omitting the guarantee of all arena work being completed.
+
+Implementation-wise, a waiting context (reference counter) could be added to the internal arena class.
+It would be incremented on each call to `execute`, `enqueue`, and `isolate`, and decremented after
+the corresponding task is complete. The more tricky part is to track parallel jobs started in an
+implicit thread-associated arenas; perhaps it can be done when registering and deregistering
+the corresponding task group contexts. For the weaker "try" semantics, it would be OK to return
+when no task to execute is found, even if some tasks are yet running; that can be done by
+inspecting the arena, without adding more reference counting.
+
+The implementation can likely be backward compatible, as no layout or function changes in `task_arena`
+seems necessary. A new library entry point would be added for the waiting function.
+
+The implementation should, if possible, assign the waiting thread to execute tasks; if not, the thread
+should *block with progress delegation*, as described below.
+
+### 3. Consider API for progress delegation
+
+Of the three options outlined above for application thread participation, the second one - exchanging
```

Comment:
Either number the options in the earlier section, or don't refer to the selected option by number at all.  It took me a minute to figure our what the three things were.  I believe those options are described in the section, "Participation of application threads" but the use of a bullet list with two items there was confusing.

Reply:
Thanks; improved.